packages/compartment-mapper/src/types/compartment-map-schema.ts

```diff
@@ -8,80 +8,231 @@
    */
   name: string;
+  modules: Record<string, T>;
+  scopes?: Record<string, ScopeDescriptor>;
+  /** language for extension */
+  parsers?: LanguageForExtension;
+  /** language for module specifier */
+  types?: LanguageForModuleSpecifier;
+  /** policy specific to compartment */
+  policy?: SomePackagePolicy;
+
   location: string;
+  /**
+   * name of the parent directory of the package from which the compartment is derived,
+   * for purposes of generating sourceURL comments that are most likely to unite with the original sources in an IDE workspace.
+   */
+  sourceDirname?: string;
+
   /**
    * whether this compartment was retained by any module in the solution. This
    * property should never appear in an archived compartment map.
    */
-  retained?: boolean;
-  modules: Record<string, ModuleDescriptor>;
-  scopes: Record<string, ScopeDescriptor>;
-  /** language for extension */
-  parsers: LanguageForExtension;
-  /** language for module specifier */
-  types: LanguageForModuleSpecifier;
-  /** policy specific to compartment */
-  policy: SomePackagePolicy;
-  /** List of compartment names this Compartment depends upon */
-  compartments: Set<string>;
-};
+  retained?: true;
+}
+
+export type CompartmentDescriptorWithPolicy<
+  T extends ModuleDescriptorConfiguration = ModuleDescriptorConfiguration,
+> = Omit<CompartmentDescriptor<T>, 'policy'> & { policy: SomePackagePolicy };
+
+/**
+ * A compartment descriptor digested by `digestCompartmentMap()`
+ */
+export interface DigestedCompartmentDescriptor
+  extends CompartmentDescriptor<ModuleDescriptorConfiguration> {
+  path: never;
+  retained: never;
+  scopes: never;
+  parsers: never;
+  types: never;
+  __createdBy: never;
+  sourceDirname: never;
+}
+
+export type DigestedCompartmentDescriptors =
+  CompartmentDescriptors<DigestedCompartmentDescriptor>;
+
+export type DigestedCompartmentMapDescriptor =
+  CompartmentMapDescriptor<DigestedCompartmentDescriptor>;
 
 /**
  * For every module explicitly mentioned in an `exports` field of a
- * `package.json`, there is a corresponding module descriptor.
+ * `package.json`, there is a corresponding `ModuleDescriptorConfiguration`.
  */
-export type ModuleDescriptor = {
-  compartment?: string;
-  module?: string;
+export type ModuleDescriptorConfiguration =
+  | SourceModuleDescriptorConfiguration
+  | CompartmentModuleDescriptorConfiguration;
+
+export type ModuleDescriptorConfigurationCreator =
+  | 'link'
+  | 'transform'
+  | 'import-hook'
+  | 'digest'
+  | 'node-modules';
+
+export interface BaseModuleDescriptorConfiguration {
+  deferredError?: string;
+
+  retained?: true;
+
+  __createdBy?: ModuleDescriptorConfigurationCreator;
+}
+
+export interface ErrorModuleDescriptorConfiguration
+  extends BaseModuleDescriptorConfiguration {
+  deferredError: string;
+}
+
+/**
+ * For every module explicitly mentioned in an `exports` field of a
+ * `package.json`, there is a corresponding `CompartmentModuleDescriptorConfiguration`.
```

Comment:
(could) 
ModuleDescriptorConfiguration and CompartmentModuleDescriptorConfiguration have the same exact description. 
It'd be nice to add the differentiating factor to the description - in what cases or why one is used and not the other.